aten/src/ATen/BatchingRegistrations.cpp

```diff
@@ -223,13 +223,33 @@ Tensor select_batching_rule(const Tensor& self, int64_t dim, int64_t index) {
   return self_physical.newLogicalFromPhysical(result);
 }
 
+static int64_t getGradInputPhysicalDim(int64_t dim, IntArrayRef input_sizes, int64_t num_batch_dims) {
+  return maybe_wrap_dim(dim, input_sizes.size()) + num_batch_dims;
+}
+
+Tensor select_backward_batching_rule(const Tensor& grad, IntArrayRef input_sizes, int64_t dim, int64_t index) {
+  auto grad_physical = MultiBatchVmapTransform::logicalToPhysical(grad);
+  auto grad_input = at::zeros(grad_physical.getPhysicalShape(input_sizes), grad.options());
+  auto physical_dim = getGradInputPhysicalDim(dim, input_sizes, grad_physical.numBatchDims());
+  grad_input.select(physical_dim, index).copy_(grad_physical.tensor());
+  return grad_physical.newLogicalFromPhysical(grad_input);
```

Comment:
I'm not sure to understand what this is doing. Where are these defined so that I can read about them?

Reply:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/VmapTransforms.h

Please feel free to ask questions about it, it is kinda confusing at times :/

Reply:
As a summary,
- auto grad_physical = MultiBatchVmapTransform::logicalToPhysical(grad);
    - This takes the logical BatchedTensor grad and turns it into a VmapPhysicalView. a VmapPhysicalView holds a regular (non-Batched) Tensor and some additional metadata. The regular (non-Batched) tensor is the grad tensor, but all of the batch/vmap dims are moved to the front of the tensor.
- auto grad_input = at::zeros(grad_physical.getPhysicalShape(input_sizes), grad.options());
    - `grad_physical` is a VmapPhysicalView. Given the input_sizes, this prepends the batch (vmap) dimensions to the input_sizes.
- auto physical_dim = getGradInputPhysicalDim(dim, input_sizes, grad_physical.numBatchDims());
    - We want to translate "dim" to the correct physical dim on the grad_input Tensor. The grad_input Tensor has some number of batch dims at the front of the tensor and then some number of regular dims.
- grad_input.select(physical_dim, index).copy_(grad_physical.tensor());
    - grad_physical.tensor() pulls out the physical Tensor from the VmapPhysicalView
- return grad_physical.newLogicalFromPhysical(grad_input);
    - Now, grad_input is a regular Tensor. newLogicalFromPhysical takes `grad_input` and turns it into a BatchedTensor by marking the first `grad_physical.numBatchDim()` dimensions as batch/vmap dimensions.

Reply:
The note at https://github.com/pytorch/pytorch/blob/a2a81e1335f28dbd21ef517f18ff218706e70ffd/aten/src/ATen/BatchingRegistrations.cpp#L8-L49 should provide some context as well

Reply:
Whether or not Alban can understand after these descriptions will be a good test, since I think this has been a me and you thing only so far XD

Reply:
I think this is quite good.
The main point of confusion for me was `newLogicalFromPhysical` that is called on a physical Tensor with a physical argument. So it is unclear from the name which one is used to create the logical output. Especially when you write `grad_physical.newLogicalFromPhysical(grad_input)` where the one with `_physical` in its name is not the one that is actually used.

I am not sure how to improve it though. It feels like there are two states in the naming right now: physical and logical. But the physical view should be there as well?
A random idea that comes to mind could be to have the 3 states be Logical, Physical and Plain? So that you have a `newLogicalFromPlain` function and `grad_physical.toPlain()` instead of `grad_physical.tensor()`.



The second point which is not really that important is that it's too bad that you cannot use `.getPhysicalShape()` here and have to re-implement it. You could maybe extend this API to have an overload that takes the plain sizes as input?

Reply:
These are good points. One thing I'd point out is that we do encode Physical and Logical distinction in the type system, so the compiler will call you out if you get it wrong. But indeed some of these improvements may be worth doing.